core/templates/services/playthrough.service.ts

```diff
@@ -17,383 +17,268 @@
+  }
 
-    private isMultipleIncorrectSubmissionsIssue(): boolean {
-      return (
-        PlaythroughService.multipleIncorrectStateName.num_times_incorrect >=
-        ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD);
-    }
+  private currentStateName(): string {
+    return this.visitedStates[this.visitedStates.length - 1];
+  }
 
-    private isCyclicStateTransitionsIssue(): boolean {
-      return PlaythroughService.cycleIdentifier.num_cycles >=
-        ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
-    }
+  isIssue(): boolean {
+    return this.numCycles >= ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
+  }
 
-    private isEarlyQuitIssue(timeSpentInExpInSecs: number): boolean {
-      return timeSpentInExpInSecs <
-        ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS;
+  recordStateTransition(destStateName: string): void {
+    if (!this.visitedStates.includes(destStateName)) {
+      this.visitedStates.push(destStateName);
+      return;
     }
-
-    private analyzePlaythrough(): void {
-      // The ordering of checks in this method is such that the priority of
-      // issues to be recorded in case of multiple issues is captured. This
-      // follows MultipleIncorrectSubmissionsIssue ->
-      // CyclicStateTransitionsIssue -> EarlyQuitIssue.
-      let timeSpentInExpInSecs = (
-        PlaythroughService.expStopwatch.getTimeInSecs());
-      if (this.isMultipleIncorrectSubmissionsIssue()) {
-        PlaythroughService.playthrough.issueType = (
-          AppConstants.ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS);
-        PlaythroughService.playthrough.issueCustomizationArgs = {
-          state_name: {
-            value: PlaythroughService.multipleIncorrectStateName.state_name
-          },
-          num_times_answered_incorrectly: {
-            value: (
-              PlaythroughService.multipleIncorrectStateName.num_times_incorrect)
-          }
-        };
-      } else if (this.isCyclicStateTransitionsIssue()) {
-        PlaythroughService.playthrough.issueType = (
-          AppConstants.ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS);
-        PlaythroughService.playthrough.issueCustomizationArgs = {
-          state_names: {
-            value: PlaythroughService.cycleIdentifier.cycle.split(',')
-          }
-        };
-      } else if (this.isEarlyQuitIssue(timeSpentInExpInSecs)) {
-        PlaythroughService.playthrough.issueType = (
-          AppConstants.ISSUE_TYPE_EARLY_QUIT);
-        PlaythroughService.playthrough.issueCustomizationArgs = {
-          state_name: {
-            value:
-              PlaythroughService.playthrough.actions[
-                PlaythroughService.playthrough.actions.length - 1
-              ].actionCustomizationArgs.state_name.value
-          },
-          time_spent_in_exp_in_secs: {
-            value: timeSpentInExpInSecs
-          }
-        };
+    if (this.currentStateName() !== destStateName) {
+      const cycle = this.makeCycle(this.visitedStates.indexOf(destStateName));
```

Comment:
If destStateName is repeated in the list, should this start from the rightmost occurrence rather than the leftmost one? (I'm not sure.)

Some of this stuff might benefit from inline comments.